src/filetree.rs

```diff
@@ -302,34 +330,85 @@ pub(crate) fn apply_diff(
     let opts = opts.unwrap_or(&default_opts);
     cleanup_tmp(destdir).context("cleaning up temporary files")?;
 
-    // Write new and changed files
-    for pathstr in diff.additions.iter().chain(diff.changes.iter()) {
-        let path = Path::new(pathstr);
-        if let Some(parent) = path.parent() {
-            destdir.ensure_dir_all(parent, DEFAULT_FILE_MODE)?;
+    let mut updates = HashMap::new();
+    // Handle removals in temp dir
+    if !opts.skip_removals {
+        for pathstr in diff.removals.iter() {
+            let path = Utf8Path::new(pathstr);
+            let (subdir, subdir_tmp) = get_subdir(path)?;
+            let path_tmp;
+            if subdir != *pathstr {
+                if !destdir.exists(&subdir_tmp)? {
+                    copy_dir(destdir, &subdir, &subdir_tmp)?;
+                    updates.insert(subdir.clone(), subdir_tmp.clone());
+                }
+                path_tmp = Utf8Path::new(&subdir_tmp).join(path.strip_prefix(&subdir)?);
+            } else {
+                // remove the file directly that is not in dir
+                path_tmp = path.to_path_buf();
+            }
+            destdir
+                .remove_file(path_tmp.as_std_path())
+                .with_context(|| format!("removing {:?}", path_tmp))?;
         }
-        let destp = tmpname_for_path(path);
-        srcdir
-            .copy_file_at(path, destdir, destp.as_path())
-            .with_context(|| format!("writing {}", &pathstr))?;
     }
     // Ensure all of the new files are written persistently to disk
     if !opts.skip_sync {
         syncfs(destdir)?;
     }
-    // Now move them all into place (TODO track interruption)
-    for path in diff.additions.iter().chain(diff.changes.iter()) {
-        let pathtmp = tmpname_for_path(path);
-        destdir
-            .local_rename(&pathtmp, path)
-            .with_context(|| format!("renaming {path}"))?;
+
+    // Write new files to temp dir if exists, else write directly in dest
```

Comment:
Theoretically, a binary could change behavior if another file is there or not.

So maybe let's move that after the changed files loop to make it safer. Or we add only to a temp dir copy in all cases.

Reply:
> So maybe let's move that after the changed files loop to make it safer. Or we add only to a temp dir copy in all cases.

Move additions after changed seems easier.
If move to temp dir, it might be different as there are no original copy and finally will use `rename()` instead of `exchange()`. We can change to temp if it is necessary. WDYT?

Reply:
Note that there will be no original copy only if the file is at the root of `EFI/`, which is rather unlikely as this is not the case in any of our distributions (but should still be handled nevertheless if we do this).

Reply:
OK, change logic here: if additional file is at the root of `EFI/` or new subdir that destination does not have, for example `testfile`, `fedora/newsub/newfile`, then write to `.btmp.testfile` , `fedora/newsub/.btmp.newfile` (reuse `TMP_PREFIX`), finally do `rename()`, is this accepted?

Reply:
> testfile -> .btmp.testfile

Agree

> fedora/newsub/newfile -> fedora/newsub/.btmp.newfile

That creates another case, and we would have to clean those files on failure. I would prefer we stick to copying the entire dir in the non-top-level case: `fedora/newsub/newfile` -> fedora.tmp/newsub/newfile` + rename_exchange.

Reply:
> copying the entire dir in the non-top-level case: `fedora/newsub/newfile` -> `fedora.tmp/newsub/newfile` + rename_exchange.

Updated, thanks!